### ISSUE TITLE
http2_protocol_options.stream_error_on_invalid_http_messaging with reverse bridge

### ISSUE THREAD

**Issue by sp-manuel-jurado** [envoyproxy/envoy#11099]
Hi,

I'm configuring envoy (v1.12.2) as level two proxy (sidecar) within reverse bridge filter enabled only, and using an ingress controller envoy based as level one proxy.

I've read [this article](https://www.envoyproxy.io/docs/envoy/latest/configuration/best_practices/level_two) and I checked using http2_protocol_options.stream_error_on_invalid_http_messaging=true, but I continue getting this error (only for grpc errors; for grpc succesful responses works correctly):
```
{
  "error": "14 UNAVAILABLE: upstream connect error or disconnect/reset before headers. reset reason: connection termination"
}
```

If I do the calls directly to the service (which contains the sidecar), both cases (grpc errors and grpc successful responses) work correctly, and there is no connection termination.

Is it possible that this config does not apply to the reverse bridge?
Would there be any workaround?

This is the filter config (the upstream doesn't understand grpc semantics):
```
- name: envoy.filters.http.grpc_http1_reverse_bridge
   config:
     content_type: application/grpc+proto
     withhold_grpc_frames: true
```

This is the http connection manager config:
```
 - name: envoy.http_connection_manager
    config:
      http2_protocol_options:
        stream_error_on_invalid_http_messaging: true
```

Thank you in advance.

**Comment by alyssawilk** [comment]
FWIW what stream_error_on_invalid_http_messaging controls is if there's an HTTP/2 level issue if the stream is closed or if the full connection is closed.  It shouldn't change the validation which is being performed, or look at the gRPC layer at all, it just controls how thoroughly HTTP/2 non-compliant traffic is rejected.

**Comment by sp-manuel-jurado** [comment]
Hi @alyssawilk ,
thanks for the quick reply.

This is a correct grpc-response (with grpc error) returned by the L2 proxy.
You can see the header "grpc-status-details":
```
Response trailers received:
access-control-allow-origin:
cache-control: no-cache, private
content-length: 5
content-type: application/grpc
date: Thu, 07 May 2020 18:41:50 GMT
grpc-status-details: {"code":2,"message":"Invalid argument type","details":{}}
server: envoy
x-envoy-upstream-healthchecked-cluster:
x-envoy-upstream-service-time: 23
Sent 1 request and received 0 responses
ERROR:
  Code: Unknown
  Message: Invalid argument type
```

This L2 proxy doesn't have any cluster/upstream that understand grpc (only http1.1), so this response is transformed by the http filter grpc-http1-reverse-bridge.

Guessing that grpc is a protocol over http2 and reading:
```
stream_error_on_invalid_http_messaging:
(bool) Allows invalid HTTP messaging and headers. When this option is disabled (default), then the whole HTTP/2 connection is terminated upon receiving invalid HEADERS frame. However, when this option is enabled, only the offending stream is terminated.
```

I suspect that envoy L1 detects "grpc-status-details" as an invalid header and close de connection, even if stream_error_on_invalid_http_messaging is set to true (due to reverse bridge casuistry). But I'm not 100% sure, maybe is another issue.

This is the L1 proxy response:
```
Request metadata to send:
(empty)

Response headers received:
(empty)

Response trailers received:
content-type: application/grpc
date: Thu, 07 May 2020 18:44:50 GMT
server: envoy
Sent 1 request and received 0 responses
ERROR:
  Code: Unavailable
  Message: upstream connect error or disconnect/reset before headers. reset reason: connection termination
```

I've checked using L1 ingress controller non envoy based and works as expected.

Note:
grpc_http1_reverse_bridge_filter:
This is a filter that enables converting an incoming gRPC request into a HTTP/1.1 request to allow a server that does not understand HTTP/2 or gRPC semantics to handle the request.

**Comment by sp-manuel-jurado** [comment]
This is the level 2 envoy configuration in case it was useful:

```
    admin:
      access_log_path: /dev/stdout
      address:
        socket_address:
          address: 127.0.0.1
          port_value: 9901
    static_resources:
      listeners:
      - address:
          socket_address:
            address: 0.0.0.0
            port_value: 10000
        filter_chains:
        - filters:
          - name: envoy.http_connection_manager
            config:
              http2_protocol_options:
                stream_error_on_invalid_http_messaging: true
              stream_idle_timeout: 40.00s
              use_remote_address: false
              access_log:
              - name: envoy.file_access_log
                config:
                  path: /dev/stdout
              stat_prefix: ingress_http_reverse
              route_config:
                name: local_route
                virtual_hosts:
                - name: local_service
                  domains: ["*"]
                  routes:
                  - match:
                      prefix: "/"
                    route:
                      host_rewrite: service-http11
                      cluster: backend
              http_filters:
                - name: envoy.filters.http.grpc_http1_reverse_bridge
                  config:
                    content_type: application/grpc+proto
                    withhold_grpc_frames: true
                - name: envoy.health_check
                  config:
                    pass_through_mode: true
                - name: envoy.router
                  config: {}
          use_proxy_proto: false
      clusters:
      - name: backend
        connect_timeout: 5.00s
        type: LOGICAL_DNS
        dns_lookup_family: V4_ONLY
        lb_policy: ROUND_ROBIN
        http_protocol_options: {}
        hosts:
          - socket_address:
              address: service-http11
              port_value: 80
```

**Comment by sp-manuel-jurado** [comment]
I add, probably, a related issue (about a year ago) in case it can give us any clue:
- https://github.com/envoyproxy/envoy/issues/6471

**Comment by alyssawilk** [comment]
Hmmm.
Well I'm unable to repro exactly - you went and made this awesome clean reproducible docker set up case and I went and stripped it down so I wasn't running anything I didn't grok (alas, security).  I snagged your go and just had it auto-return error, and ran your configs locally sans docker
../bazel-bin/source/exe/envoy-static --config-path  ./envoy_l1.yaml -l trace
../bazel-bin/source/exe/envoy-static --config-path  ./envoy_l2.yaml -l trace --base-id 3
(with the yamls tweaked to hit up 127.0.0.1)

I don't have grpc-curl, but using telnet and hitting up telnet localhost 11001
GET /sp.rpc.hello.HelloService/Hello HTTP/1.1
host: asd

and what I get is the L2 proxying the error to the L1, and the L1 proxying valid HTTP/1 (possibly not Ok gRPC) to the client

HTTP/1.1 500 Internal Server Error
content-length: 0
content-type: application/grpc+proto
grpc-message: invalid name
grpc-status: 3
date: Mon, 11 May 2020 19:22:29 GMT
x-envoy-upstream-service-time: 24
server: envoy

it's sending end stream so there's something subtly different between what you're doing and what I'm doing.  

If you are comfortable building and running envoy from head I'm wondering if the difference persists?  Alternately if you can run your envoys with -l trace and send me the logs, I can try to see what's going on on your side, but if it's working at HEAD my advice would be to upgrade :-)

**Comment by sp-manuel-jurado** [comment]
Hi @alyssawilk,

here the traces: https://gist.github.com/sp-manuel-jurado/5b0bc9cbf630685cf46d05a1a1411ef0

- For envoy_l1 I've enabled `-l trace`
- For envoy_l2 I've enabled `-l trace --base-id 3`
- https://github.com/sp-manuel-jurado/envoyproxy-envoy-issues-11099/commit/332d5a2750f693168bb16b400ec76a62d348cf9a

I checked using `envoy-alpine-debug-dev:bcc3050f7385a3f67c1b62b8779c44c7cece6b10` https://github.com/envoyproxy/envoy/commit/bcc3050f7385a3f67c1b62b8779c44c7cece6b10 and It seems continues failing :(
```
rpc error: code = Unavailable desc = upstream connect error or disconnect/reset before headers. reset reason: connection termination
```

> it's sending end stream so there's something subtly different between what you're doing and what I'm doing.
- probably you're doing calls in http1.1 and in my case, I'm doing calls in gRPC/HTTP2

Notes:
- grpcurl: https://github.com/fullstorydev/grpcurl

**Comment by alyssawilk** [comment]
OK, I think I see the problem.  

I'm seeing the router receiving headers with content-length: 5 but end_stream=true
then your L1 resets due to messaging violation (my guess due to that content length not matching end stream) which is a reasonable thing.

It looks like on the response path due to the withheld frames, the content length is adjusted to accommodate the frames, but because it's a header-only response the headers are encoded with content-length: 5 but end_stream = true.   The data is not appended because there's no onData call.  I think the bridge filter simply doesn't handle header-only responses correctly. 

tagging @snowp and @zuercher as codeowners - sending a content length with no content isn't right but I'm not 100% sure what the correct change is.

If it's as simple as changing encodeHeaders(...) {
if (withhold_grpc_frames_) {
to 
if (withhold_grpc_frames_ && !end_stream) {

please throw it my way.  If the grpc frames still need to be echod back I'll let someone with more familiarity take a look

**Comment by snowp** [comment]
I think that fix should do it, the code just blindly assumed that there would be a payload in the response. Another way of doing it would be to have the content-length adjustment do nothing if the upstream content-length is 0.

**Comment by alyssawilk** [comment]
Cool, will have a fix out shortly.
In the meantime @sp-manuel-jurado I think  you may be able to work around this bug by simply not sending the zero content length header (if your go server allows it) since we only adjust the content length if the header is present.

**Comment by sp-manuel-jurado** [comment]
> Cool, will have a fix out shortly.
> In the meantime @sp-manuel-jurado I think you may be able to work around this bug by simply not sending the zero content length header (if your go server allows it) since we only adjust the content length if the header is present.

Thx @alyssawilk for finding the bug.
Related to workaround, unfortunately, go server, as the standard way, (It seems) even if you delete content-length: 0 header it injects it automatically (in case there is no content).
Also, I did the workaround in PHP. It allows to delete de content-length header. There is no longer: "connection termination" but grpc-message header seems that It isn't propagated correctly by reverse-bridge filter.

Here de response (you can see empty ERROR.Message at the end):
```
Response headers received:
access-control-allow-origin:
cache-control: no-cache, private
content-type: application/grpc
date: Wed, 13 May 2020 11:10:11 GMT
grpc-status-details: {}
server: envoy
x-envoy-upstream-healthchecked-cluster:
x-envoy-upstream-service-time: 3703
x-powered-by: PHP/7.3.5

Response trailers received:
(empty)
Sent 0 requests and received 0 responses
ERROR:
  Code: Unknown
  Message:
```

So I'll wait to check it again (using master dev image) when the fix (https://github.com/envoyproxy/envoy/pull/11166) is merged.

**Comment by alyssawilk** [comment]
Ok, that's just whacky.  The only place I see removeGrpcMessage in the code base is the grpc_json_transcoder which you're not using.  
We set it in 4 places - where we send a local reply (this is clearly a proxied response given the powered by PHP), the bridge and transcoder filters (not using) and in the reverse bridge filter iff the content type is bad, but then it'd be present not reomved.  Are you sure the upstream is sending them correctly?  Again if you throw me -l trace files I can check it out.

**Comment by sp-manuel-jurado** [comment]
Hi @alyssawilk,

yes upstream in http1.1 is returning grpc-message header within "Invalid argument type".

Here the new traces (traces_02.log), only for rpc call to proxy level 1 (and removing `'Content-Length' => 0` header in upstream):
- https://gist.github.com/sp-manuel-jurado/daae663d057523c3342359851c57d515
(for security reasons I changed the rpc call info in the traces by rpc.TestService/Test)

Here the grpcurl response:
```
Response headers received:
access-control-allow-origin:
cache-control: no-cache, private
content-type: application/grpc
date: Wed, 13 May 2020 15:15:02 GMT
grpc-status-details: {"...":"..."}
server: envoy
x-envoy-upstream-healthchecked-cluster:
x-envoy-upstream-service-time: 6579
x-powered-by: PHP/7.3.5

Response trailers received:
(empty)
Sent 1 request and received 0 responses
ERROR:
  Code: Unknown
  Message:
```

- "envoy_l1" is L1 proxy docker.io/envoyproxy/envoy-alpine-debug-dev:bcc3050f7385a3f67c1b62b8779c44c7cece6b10 with -l trace (listening 11001)
- "envoy" is L2 proxy docker.io/envoyproxy/envoy-alpine-debug-dev:bcc3050f7385a3f67c1b62b8779c44c7cece6b10 (-l trace --base-id 3)

**Comment by alyssawilk** [comment]
Ok, in this case looking here:
https://gist.github.com/sp-manuel-jurado/daae663d057523c3342359851c57d515#file-traces_02-log-L678

We're sending the full request headers, but it's not a header-only response.
I think in this case we hit the thing @snow and I discussed where we end up serializing an empty gRPC response (fine) and then synthesize trailers.  The trailers only capture the grpc-status from the headers, so you can see the trailers showing being forwarded only with the status

I think it would make sense to forward on the grpc-message in the trailers, as that's a header Envoy supports.  Given your client is still saying error code unknown though, and it looks to me like the L1 Envoy is sending it

https://gist.github.com/sp-manuel-jurado/daae663d057523c3342359851c57d515#file-traces_02-log-L887

I think there may be some disconnect between what Envoy is sending and what grpc-curl is expecting.  Perhaps it expects the proto from the grpc-status-details to be forward as grpc response body? Might be worth looping in @jhump or other grpc-curl folks for what that would expect to see.

I suspect the content-length 0 fix should help you given that go was doing a header-only response but I suspect PHP won't work and I'm unclear if forwarding on the details would be sufficient.

**Comment by sp-manuel-jurado** [comment]
Related to `grpc-status-details` headers:
If a delete `Content-Type:0` and `grpc-status-details` (workaround) from http1.1 upstream (PHP)... same behavior occurs, nothing seems to change.

upstream http1.1 response (curl):
```
< HTTP/1.1 500 Internal Server Error
< Server: nginx/1.16.0
< Content-Type: application/grpc+proto
< Transfer-Encoding: chunked
< Connection: keep-alive
< X-Powered-By: PHP/7.3.5
< grpc-status: 2
< grpc-message: Invalid argument type
< Cache-Control: no-cache, private
< Date: Thu, 14 May 2020 15:14:18 GMT
< Access-Control-Allow-Origin:
<
```

envoy L2 response (grpcurl):
```
Response headers received:
access-control-allow-origin:
cache-control: no-cache, private
content-type: application/grpc
date: Thu, 14 May 2020 15:13:08 GMT
server: envoy
x-envoy-upstream-healthchecked-cluster:
x-envoy-upstream-service-time: 1686
x-powered-by: PHP/7.3.5

Response trailers received:
(empty)
Sent 1 request and received 0 responses
ERROR:
  Code: Unknown
  Message:
```

envoy L1 response (grpcurl):
```
Response headers received:
access-control-allow-origin:
cache-control: no-cache, private
content-type: application/grpc
date: Thu, 14 May 2020 15:13:20 GMT
server: envoy
x-envoy-upstream-healthchecked-cluster:
x-envoy-upstream-service-time: 2509
x-powered-by: PHP/7.3.5

Response trailers received:
(empty)
Sent 1 request and received 0 responses
ERROR:
  Code: Unknown
  Message:
```

Once fix https://github.com/envoyproxy/envoy/pull/11166 is merged I can check again.

**Comment by jhump** [comment]
I'm not entirely sure I understand the issue. But do note that `grpcurl` just uses the Go gRPC runtime library for all of the connection and protocol handling. The latest release is using gRPC Go 1.28, but I am actually prep'ing a release to update that to latest (1.29).

**Comment by alyssawilk** [comment]
FWIW I think the PR fixed the specific issue you were encountering, but it sounds like there's possibly more than one :-/
Try it out and LMK how it goes.  Again not my area of code but I'll still try to take a whack at it when I have cycles :-)
